# Mini-project part III: Research report

We've put some effort into building our collection of volumes - you can find them in [this Google Drive folder](https://drive.google.com/drive/u/0/folders/1UAaGIiqElF9YLTGIy6hQYM7QcGzosZgR). Now it's time to learn something about it. You already have lots of excellent ideas for how to apply the tools we've learned about so far. It's also a good time in the semester to review what we have learned and practice applying it in less structured settings.

**You will work by yourself or in a group of up to three people** to complete a short project applying methods from the previous weeks to this collection. You will turn in the completed project as a single notebook (one submission per group) with the following sections:

0. **Project team.** List members with full names and NetIDs. If your group does not contain at least one native speaker of English, let us know.

1. **Question(s) (10 points).** Describe what you wanted to learn. Suggest several possible answers or hypotheses, and describe in general terms what you might expect to see if each of these answers were true (save specific measurements for the next section). For example, many students want to know the difference between horror and non-horror fiction, or between detective stories and horror fiction, but there are many ways to operationalize this question. You do not need to limit yourself to questions of genre. **Note that your question should be interesting! If the answer is obvious before you begin, or if it's something the importance of which you cannot explain, your grade will suffer (a lot).** 

1. **Methods (10 points).** Describe how you will use computational methods presented so far in this class to answer your question. What do the computational tools do, and how does their output relate to your question? Describe how you will process the collection into a form suitable for a model or algorithm and why you have processed it the way you have.

1. **Code (20 points).** Carry out your experiments. Code should be correct (no errors) and focused (unneeded code from examples is removed). Use the notebook format effectively: code may be incorporated into multiple sections.

1. **Results and discussion (40 points).** Use sorted lists, tables, and visual presentations to make your argument. Excellent projects will provide multiple views of results, and follow up on any apparent outliers or strange cases, including through careful reading of the original documents.

1. **Reflection (10 points).** Describe your experience in this process. What was harder or easier than you expected? What compromises or negotiations did you have to accept to match the collection, the question, and the methods? What would you try next? Your reflections should be written as single narrative that incorporates the viewpoints of all group members.

1. **Resources consulted (0 points, but -5 if missing).** Credit any online sources (Stack Overflow, blog posts, documentation, etc.) that you found helpful.

1. **Responsibility statement (0 points, but -5 if missing).** See separate CMS assignment "MP 03: Responsibility statement". **Note:** If you worked alone on the project (a group of one), you are not required to submit the responsibility statement.


## Submission instructions

1. Complete this Jupyter Notebook.
1. Open a group submission on CMS - Let us know if you encounter any problems with that!
1. Assign the role of group submitter to a member of your group.
1. Submit the completed Jupyter Notebook in the group submission on CMS.

**Note:** If you worked alone on the project (a group of one), you are not required to submit the responsibility statement.

Otherwise...

1. Go to the CMS assignment 'MP 03: Responsibility statement'.
1. Complete the statement, describing each group member's contributions as you see them. 
1. Upload the statement as an **individual** submission on CMS.


## Guidance and advice

Show us what you've learned so far. Try to use a range of methods while remaining focused on your chosen problem. For inspiration, consider the range of research problems you've enoucountered in the readings. Note that dictionary-based sentiment scoring projects have not historically done well without substantial additional methodological diversity.

**We will grade this work based on accuracy, thoroughness, creativity, reflectiveness, and quality of presentation.** Code and results that are merely correct are generally not enough, on their own, to achieve a high score.

**Scope:** this is a *mini*-project, with a short deadline. We are expecting work that is consistent with that timeframe, but that is serious, thoughtful, and rigorous. This assignment will almost certainly require more time and effort than the typical weekly homework. **For group work, the expected scope grows linearly with the number of participants.**

# 0. Project team

Estelle Hooper (ehh52), Gabriella Chu (gc386)

# 1. Question(s) (10 points)

### Research Question: How can we recommend books of other genres to lovers of a single genre?
#### Motivation 
When choosing to consume any form of entertainment, such as a TV show, play, or movie, people tend to favor a particular genre or gravitate towards a favorite story. Unlike the aforementioned examples, people tend to consume books or novels at their own pace, which could be considered more time consuming and costly in terms of mental effort to stay committed to a written story. Because time and attention are scarce, people are more likely to read books of one genre. Many people do not have the capacity to branch out and try new genres, feeling that the safer choice is to read a more predictable book of their favored genre rather than risk not enjoying content of a different genre.
#### Hypothesis
In order to have confidence in another genre, there must be overlapping qualities between a reader's favorite genre and the new genre. Fans of the horror genre may be fond of the setting because horror fiction can often involve supernatural beings or powers, such as ghosts. Therefore, they may also enjoy science fiction or fantasy that similarly incorporates elements outside of the natural world. Besides content, genres may overlap in how the genre's authors tend to portray characters. Adventure and detective genres tend to involve a partner-in-crime in which the two characters develop strong trust. If the character interactions are important, then perhaps fans of the detective genre would also be a fan of the romance genre since the latter involves the development of a relationship between characters.
#### Project
Our corpus consists of a good sample of horror novels, detective novels, and a miscellanous group of novels from other genres. We want to see how close these "other novels" are to either the horror genre or detective genre. We will train a classifier using the detective and horror novels, and run that classifier on the "other" novels. Those "other" novels will receive a horror or detective label (despite not being canonically considered part of the horror or detective genres). With these classifier outcomes, we will recommend the "other" books labeled "detective" to fans of the detective genre and those labeled "horror" to the horror genre. For example, the classifier ran on *Harry Potter*, a fantasy series, and gave it a detective label, we would recommend detective fans to read *Harry Potter.*
#### Expectations
Overall, we don't expect the recommendations we obtain from the classifer to be very great because it is only working with detective and horror, and it will be hard to find overlapping words with only those two genres and the various other genres in the corpus. We expect detective and horror novels with the most subplots and settings similar to other genres to score well with those genres. For example, if detective novels tend to have romance subplots, we expect the classifier to label romance novels as detective.

# 2. Methods (10 points)

### I. Data Preparation
From the class corpus, we will divide the novels into two subsets: horror and detective, and the rest of the novels. The horror and detective will act as our "training" dataset because we will train the classifier to classify text as either **detective (==1) or horror (==0).** All the other novels of other genres will act as our "testing" dataset, and the classifier will predict these novels as either horror or detective. We will use the detective column in the metadata to get the gold labels of the training data in order to score our model later on. We'll also read/open all the novels in this step.
### II. Vectorization
We will vectorize the novels with tfidf weighting, L2, and z-scores to standardize and normalize our corpus, and consider pre-processing factors like stopwords and lowercasing. Them, we'll create multilple vectorizers with a variety of max input features and fit the training data (true horror and detective). We will try different combinations of these matrices (each with a different number input features) with different types of classifiers in the next step.
### III. Classifier
We will test different classification methods with different parameters (Multinomial Naive Bayes and Logistic Regression) with our X_train matrix of different sizes and the gold labels (y_train) and compare the accuracy/precision/recall/F1 scores and select the best classifier.
### IV. Feature Importance 
After finding the model with the best F1 score, we will examine the words/tokens/features with the heaviest weights that aid in classifying a work as detective or horror. If we see odd words, such as character names, or common stopwords as the top features, we will reconsider our vectorization steps.
### V. Testing/Predicting
After finalizing our classifier, we will run it on our testing data, or the novels of other genres without detective or horror labels. From this classifier, each "other" novel will receive a horror or detective label (y_test).
### VI. Scatterplots/Visualizations (to be discussed in Part 4: Results)
Because the classifier only makes binary decisions, we want to examine the most "detective-y" and "horror-y" novels because there will be some novels that the classifer could not make as definitive decisions for. We will use SVD to graph all the novels in 2D space. We will then create multiple scatterplots to look at these decision boundaries. The most important scatterplots will be the ones that look at how close the true labeled novels are to the predicted  novels are in distance space. We will consider the predicted dots that are closest to the true dots to be the best recommendations for that genre.

# 3. Code (20 points)

In [42]:
# Imports (all of them!)
%matplotlib inline
import string
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import os
import seaborn as sns
from   sklearn.decomposition import TruncatedSVD
from   sklearn.feature_extraction.text import TfidfVectorizer
from   sklearn.preprocessing import StandardScaler, MinMaxScaler, normalize
from   sklearn.metrics.pairwise import euclidean_distances
from sklearn.model_selection import cross_validate
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import MultinomialNB

## I. Data Preparation
In this section, we cleaned and manipulated the class metadata in order to subset the data into training and testing. As stated previously, our training data is all horror or detective, and our testing is all other books that are not horror and detective. The following is a list of variables that we created to accomplish in order to proceed with vectorization. Besides subsetting, the most notable tasks we did were sorting the dataframes by title in alphabetical order and obtaining the gold labels (y_train), which will be used to cross_validate our model when we build our classifer. 
- `metadata` df; metadata spreadsheet from class corpus
- `training_data` df; metadata with training novels, received by subsetting metadata Detective==True | Horror==True
- `testing_data` df; metadata with testing novels, received by subsetting metadata Detective==True & Horror==True
- `training_names` list; list of filenames, recieved by training_data.filename.values
- `testing_names` list; list of filenames, recieved by testing_data.filename.values
- `training_books` list; list of strings/books, recieved by reading files using training_names
- `testing_books` list; list of strings/books, recieved by reading files using testing_names
- `y_train` list; the **gold labels (1=detective, 0=horror)**, or true values of the training data, obtained from the training_data detective column


In [43]:
# class corpus metadata
metadata = pd.read_csv("class_corpus_metadata.csv")
metadata.shape

(160, 34)

In [44]:
metadata.head()

Unnamed: 0 check_1 check_2  \
0      nsg57  scw222   lcc82   
1      fhh26   gs542   tj256   
2     cl2264     NaN     NaN   
3      ehh52  sjr255   kg428   
4      dgr73  jlp367   kg428   

                                               title  year author1_surname  \
0        Writings in the United Amateur, 1915 - 1922  1922       Lovecraft   
1                                        Whose Body?  1923          Sayers   
2                                      Voodoo Planet  1959          Norton   
3  Varney the Vampire; Or, the Feast of Blood by ...  1845           Rymer   
4                                  Uncle Tom's Cabin  1852           Stowe   

  author1_givenname author2_surname author2_givenname gender_author1  ...  \
0            Howard             NaN               NaN           Male  ...   
1        Dorothy L.             NaN               NaN         Female  ...   
2             Andre             NaN               NaN         Female  ...   
3     James Malcolm           Prest    Thomas Peckett           Male  ...   
4   Harriet Beecher             NaN               NaN         Female  ...   

  feminist fiction mystery adventure tragedy children  regency  manners  \
0            False    True     False   False    False    False    False   
1            False    True     False   False    False    False    False   
2            False   False      True   False    False    False    False   
3            False   False     False   False    False    False    False   
4            False   False     False   False    False    False    False   

  philosophical coming-of-age  \
0          True         False   
1         False         False   
2         False         False   
3         False         False   
4         False         False   

                                            filename  
0  Lovecraft_WritingsintheUnitedAmateur1915-1922.txt  
1                               Sayres_WhoseBody.txt  
2                            Norton_VoodooPlanet.txt  
3                  Prest_Rhymer_VarneyTheVampire.txt  
4                          Stowe_UncleTom_sCabin.txt  

[5 rows x 34 columns]

In [45]:
# training data are books that are either horror or detective
training_data = metadata[(metadata['horror']==True) | (metadata['detective']==True)]

# drop books that are both horror and detective
drop = metadata[(metadata['horror']==True) & (metadata['detective']==True)]
training_data = training_data.drop(drop.index)

# testing data are books are neither horror or detective
testing_data = metadata[(metadata['horror']==False) & (metadata['detective']==False)]

# sort titles alphabetically 
training_data = training_data.sort_values('title')
testing_data = testing_data.sort_values('title')
# note: training+testing+dropped row = 159 rows, class corpus = 160 rows, "An Unkindness of Ghosts" has no input for horror and detective column

### There are 80 combined horror and detective novels in the corpus that we will use to train the classifier.

In [46]:
training_data=training_data.reset_index(drop=True)
training_data

Unnamed: 0 check_1 check_2  \
0       tl566   hz542   ja532   
1       gc386     NaN     NaN   
2       nca28   tl566   stw43   
3      jc2739     NaN     NaN   
4       lcc82   yk499     NaN   
..        ...     ...     ...   
75      tr333  sjs457  sl2324   
76     lrs263   sh785   hz542   
77      ehh52  sjr255   kg428   
78      fhh26   gs542   tj256   
79      nsg57  scw222   lcc82   

                                                title  year author1_surname  \
0                                                 813  1910         Leblanc   
1                             A Strange Disappearance  1998           Green   
2                                  A Study in Scarlet  1887     Conan Doyle   
3                                         Agatha Webb  1899           Green   
4                                            Carmilla  1872         Le_Fanu   
..                                                ...   ...             ...   
75                                 The Valley of Fear  1915           Doyle   
76                         The Wisdom of Father Brown  1914      Chesterton   
77  Varney the Vampire; Or, the Feast of Blood by ...  1845           Rymer   
78                                        Whose Body?  1923          Sayers   
79        Writings in the United Amateur, 1915 - 1922  1922       Lovecraft   

   author1_givenname author2_surname author2_givenname gender_author1  ...  \
0            Maurice             NaN               NaN           Male  ...   
1     Anna Katharine             NaN               NaN         Female  ...   
2             Arthur             NaN               NaN           Male  ...   
3     Anna Katharine             NaN               NaN         Female  ...   
4    Joseph Sheridan             NaN               NaN           Male  ...   
..               ...             ...               ...            ...  ...   
75      Arthur Conan             NaN               NaN           Male  ...   
76     Gilbert Keith             NaN               NaN           Male  ...   
77     James Malcolm           Prest    Thomas Peckett           Male  ...   
78        Dorothy L.             NaN               NaN         Female  ...   
79            Howard             NaN               NaN           Male  ...   

   feminist fiction mystery adventure tragedy children  regency  manners  \
0             False    True     False   False    False    False    False   
1             False    True     False   False    False    False    False   
2             False    True     False   False    False    False    False   
3             False    True     False   False    False    False    False   
4             False   False     False   False    False    False    False   
..              ...     ...       ...     ...      ...      ...      ...   
75            False   False     False   False    False    False    False   
76            False   False     False   False    False    False    False   
77            False   False     False   False    False    False    False   
78            False    True     False   False    False    False    False   
79            False    True     False   False    False    False    False   

   philosophical coming-of-age  \
0          False         False   
1          False         False   
2          False         False   
3          False         False   
4          False         False   
..           ...           ...   
75         False         False   
76         False         False   
77         False         False   
78         False         False   
79          True         False   

                                             filename  
0                                     Leblanc_813.txt  
1        GreenAnnaKatharine_AStrangeDisappearance.txt  
2                      ConanDoyle_AStudyInScarlet.txt  
3                                Green_AgathaWebb.txt  
4                                        Carmilla.txt  
..                                                ...  
75    

### There are 78 combined novels from a variety of genres that are not horror or detective in the corpus.

In [47]:
testing_data=testing_data.reset_index(drop=True)
testing_data

Unnamed: 0 check_1 check_2                                  title  year  \
0       tr333  sjs457  sl2324                          A Round Dozen  1883   
1        kwy3   cl922   hk627                    A Sicillian Romance  1790   
2        lqz4   gt294   lcc82        Adele Doring at Boarding-School  1921   
3      yc2669    xf89   wms87                             Agnes Grey  1847   
4       mn454  ar2465  jlp367                  An Old-Fashioned Girl  1869   
..        ...     ...     ...                                    ...   ...   
73     jc2739     NaN     NaN                  This Side of Paradise  1920   
74      vs339   thh55     NaN                  To Kill A Mockingbird  1960   
75      fhh26   gs542   tj256  Twenty Thousand Leagues Under the Sea  1870   
76      dgr73  jlp367   kg428                      Uncle Tom's Cabin  1852   
77     cl2264     NaN     NaN                          Voodoo Planet  1959   

   author1_surname author1_givenname author2_surname author2_givenname  \
0         Coolidge             Susan             NaN               NaN   
1        Radcliffe          Ann Ward             NaN               NaN   
2            North         Grace May             NaN               NaN   
3           Bronte              Anne             NaN               NaN   
4           Alcott        Louisa May             NaN               NaN   
..             ...               ...             ...               ...   
73      Fitzgerald          F. Scott             NaN               NaN   
74             Lee            Harper             NaN               NaN   
75           Verne             Jules             NaN               NaN   
76           Stowe   Harriet Beecher             NaN               NaN   
77          Norton             Andre             NaN               NaN   

   gender_author1  ... feminist fiction mystery adventure tragedy children  \
0          Female  ...            False   False     False   False    False   
1          Female  ...            False   False     False   False    False   
2          Female  ...            False   False     False   False     True   
3          Female  ...             True   False     False   False    False   
4          Female  ...            False   False     False   False     True   
..            ...  ...              ...     ...       ...     ...      ...   
73           Male  ...            False   False     False   False    False   
74         Female  ...            False   False     False   False    False   
75           Male  ...            False   False      True   False    False   
76         Female  ...            False   False     False   False    False   
77         Female  ...            False   False      True   False    False   

    regency  manners philosophical coming-of-age  \
0     False    False         False         False   
1     False    False         False         False   
2     False    False         False         False   
3     False     True         False         False   
4     False     True         False          True   
..      ...      ...           ...           ...   
73    False    False         False          True   
74    False    False         False         False   
75    False    False         False         False   
76    False    False         False         False   
77    False    False         False         False   

                                filename  
0               Coolidge_ARoundDozen.txt  
1   radcliffeann_a_sicillian_romance.txt  
2       adele_doring_boarding_school.txt  
3                   Bronte_AgnesGrey.txt  
4         Alcott_AnOld-FashionedGirl.txt  
..                                   ...  
73     Fitzgerald_ThisSideOfParadise.txt  
74            Lee_ToKillAMockingbird.txt  
75       Verne_TwentyThousandLeagues.txt  
76             Stowe_UncleTom_sCabin.txt  
77               Norton_VoodooPlanet.txt  

[78 rows x 34 columns]

### Opening book files

In [48]:
# get book file names to open
training_names = training_data.filename.values
testing_names = testing_data.filename.values

In [49]:
print('First book in the training dataset:',training_names[0])
print('First book in the testing dataset:',testing_names[0])

First book in the training dataset: Leblanc_813.txt
First book in the testing dataset: Coolidge_ARoundDozen.txt


In [50]:
# 1=detective, 0=horror， gold labels
y_train=(training_data.detective.values*1).astype('int')
y_train

array([1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0])

In [51]:
# open and append training books together
training_books=[]
for book in training_names:
    with open(book, 'r',encoding='utf-8') as f:
        file = f.read().replace("\n", " ") 
        training_books.append(file)
# open and append testing books together
testing_books=[]
for book in testing_names:
    with open(book, 'r',encoding='utf-8') as f:
        file = f.read().replace("\n", " ") 
        testing_books.append(file)

## II. Vectorization, stopwords, normalization, standardization, matrix fitting

### Stopwords
We played around with using stopwords and changing the max_df parameter in the TfidfVectorizer(). We initially did use punctuation as stopwords and compared the matrix with and without using that parameter, and the resulting shape was the same. Given that result, we decided to ignore the stop_words parameter and let the idf weighting take care of the normalization, in addition to L2. We picked L2 normalization over L1 because it L2 downweights the lower features, preserving sparsity, using euclidean distances and helping the explainability of our model which takes in a large about of features across many genres. The max_df =.8 also takes care of the issue of stopwords since it does not consider tokens that are in more than 80% of the corpus.

In [52]:
punct=[]
for x in string.punctuation:
    punct.append(x)
punct.append('--')
punct.append('`')
punct.append("“")
punct.append("”")

### Vectorization (+ normalization)

We lowercased the words in the books and set up a vectorizer without setting up the max_features parameter. The resulting matrix shows that the training dataset of 80 true detective or  horror novels has a matrix shape of 30,376 features. We also made sure to normalize the matrix because the word count for novels can range from a minimum of 20,000 words to above 100,000 words.

In [53]:
# Custom preprocessing to remove escaped characters in input, taken from MP02
def pre_proc(x):
    '''
    Takes a unicode string.
    Lowercases, strips accents, and removes some escapes.
    Returns a standardized version of the string.
    '''
    import unicodedata
    return unicodedata.normalize('NFKD', x.replace("\'", "'").replace("\ in\ form", " inform").lower().strip())

# Set up vectorizer

vectorizer = TfidfVectorizer(
    encoding='utf-8',
    preprocessor=pre_proc,
   # stop_words=punct,
    min_df=2, # Note this
    max_df=0.8, # This, too
    binary=False,
    norm='l2',
    use_idf=True, # And this,
    #max_features=10000
)

# Your code here
X_train = vectorizer.fit_transform(training_books)
print("Matrix shape:", X_train.shape)

Matrix shape: (80, 30376)


### Standardization: Z-scores

Because the training data set consists of two genres, and the testing data set consists of multiple genres, we found it especially important to standardize the matrix with Z-scores to make features that seemed important in one genre also relevant across the other genre. It is important to note that because the matrix (X_train) is sparse, we had to set with_mean=False because the function would raise an error. This means that our mean is not scaled to 0, but rather .292. We posted on EdStem privately about this issue more than 24 hrs but were not able to get a response. 

In [54]:
#standardization
X_train_Z = StandardScaler(with_mean=False).fit_transform(X_train)
display(X_train_Z)
print('z-scored l2 mean:', round(np.mean(X_train_Z),3))
#np.std(X_train)

<80x30376 sparse matrix of type '<class 'numpy.float64'>'
	with 376031 stored elements in Compressed Sparse Row format>

z-scored l2 mean: 0.292


### Matrix fitting

We want to vectorize the corpus with a variety of different inputs to place into the classifer in addition to the matrix with all the features at 30,376. As seen below, we chose a range of input numbers beginning at 5,000 to 30,376 and saved it to a dictionary `matrix_n` to be ran in combination with the classifiers. 

In [55]:
vect_n=[]
matrix_n= {}
feat_n = [5000,10000,15000,17500,20000,22500,25000,30000,30376] # NOTE, get all the features we had to use a value greater than the number of features. So 35000 is actually just all features
for x in feat_n:
    vectorizer = TfidfVectorizer(
        encoding='utf-8',
        preprocessor=pre_proc,
        min_df=2, # Note this
        max_df=0.8, # This, too
        binary=False,
        norm='l2',
        use_idf=True, # And this
        max_features=x)
    vect_n.append(vectorizer)
    matrix = vectorizer.fit_transform(training_books)
    X_train_Z = StandardScaler(with_mean=False).fit_transform(matrix)
    dict_key=str(x)
    matrix_n[dict_key] = X_train_Z

In [56]:
matrix_n

{'5000': <80x5000 sparse matrix of type '<class 'numpy.float64'>'
 	with 187794 stored elements in Compressed Sparse Row format>,
 '10000': <80x10000 sparse matrix of type '<class 'numpy.float64'>'
 	with 274957 stored elements in Compressed Sparse Row format>,
 '15000': <80x15000 sparse matrix of type '<class 'numpy.float64'>'
 	with 320917 stored elements in Compressed Sparse Row format>,
 '17500': <80x17500 sparse matrix of type '<class 'numpy.float64'>'
 	with 335934 stored elements in Compressed Sparse Row format>,
 '20000': <80x20000 sparse matrix of type '<class 'numpy.float64'>'
 	with 347813 stored elements in Compressed Sparse Row format>,
 '22500': <80x22500 sparse matrix of type '<class 'numpy.float64'>'
 	with 357000 stored elements in Compressed Sparse Row format>,
 '25000': <80x25000 sparse matrix of type '<class 'numpy.float64'>'
 	with 364318 stored elements in Compressed Sparse Row format>,
 '30000': <80x30000 sparse matrix of type '<class 'numpy.float64'>'
 	with 375

## III. Classifier

The classifiers we were comfortable in interpreting and using for text classification were logistic regression, decision trees, random forest, multinomial naive bayes, and k-nearest neighbors. We decided to not go with Decision Trees because they are prone to overfitting and K-nearest neighbors because it is "not usually best for real-world cases" (Lec 12). We were considering going with random forest, but found it too computationally expensive. We decided on logistic regression and naive bayes because they were interpretable, not too computationally expensive, and perform relatively well.

**After testing these classifiers, we found that with cross validation the classifier that performed the best on our training data is MultinomialNB() with the default parameters, using a matrix that was vectorized with max_features=17,500 (process shown below).**

In [57]:
# Examine the performance of our simple classifiers
# Freebie function to summarize and display classifier scores
# source: mp2
def compare_scores(scores_dict):
    '''
    Takes a dictionary of cross_validate scores.
    Returns a color-coded Pandas dataframe that summarizes those scores.
    '''
    df = pd.DataFrame(scores_dict).T.applymap(np.mean).style.background_gradient(cmap='RdYlGn')
    return df

### Multinomial Naive Bayes

We played around with different parameters for the multinomial naive bayes classifier but noticed that there was no difference between using fit_prior and not using it. The MultinomialNB() classifier also has a class_prior parameter to indicate the ratio of the labels, but we decided not to use it because we are testing on a dataset that is not horror or detective, so the weights do not apply.

In [58]:
nb_classifiers = {
    'M NB Default, Alpha=1':MultinomialNB(alpha = 1),
    'M NB fit_prior=False':MultinomialNB(fit_prior = False),
}
scores = {} # Store cross-validation results in a dictionary
for classifier in nb_classifiers: 
    scores[classifier] = cross_validate( # perform cross-validation
        nb_classifiers[classifier], # classifier object
        X_train_Z, # feature matrix
        y_train, # gold labels
        cv=10, #number of folds
        scoring=['accuracy','precision', 'recall', 'f1', 'f1_macro', 'f1_micro'] # scoring methods
    )
       
compare_scores(scores)

#### Conclusion of testing different MultiNomialNB() parameters: 
Even though we tried different parameters it did not have an effect on the outcome of any of the scoring for the classifier, so we stuck with the default, and will use the default parameters to over the different sized matrices.

In [59]:
scores = {} # Store cross-validation results in a dictionary
for matrix in matrix_n: 
    print(matrix_n[matrix].shape)
    scores[matrix] = cross_validate( # perform cross-validation
        MultinomialNB(alpha = 1), # classifier object
        matrix_n[matrix], # feature matrix
        y_train, # gold labels
        cv=10, #number of folds
        scoring=['accuracy','precision', 'recall', 'f1', 'f1_macro', 'f1_micro'] # scoring methods
    )

(80, 5000)
(80, 10000)
(80, 15000)
(80, 17500)
(80, 20000)
(80, 22500)
(80, 25000)
(80, 30000)
(80, 30376)


In [60]:
compare_scores(scores)

#### Conclusion of testing MultiNomialNB() over different sized inputs: 
The default MultinomialNB() with a matrix of max_features=17500 performed the best across all scoring, with an F1 score at .9303. We decided to toy with a more specified range around  17500 features to try to refine this score (in the following cell). 

In [61]:
vect_n2=[]
matrix_n2= {}
feat_n2 = [15000,16000,17000,18000,19000,20000,17500]
for x in feat_n2:
    vectorizer = TfidfVectorizer(
        encoding='utf-8',
        preprocessor=pre_proc,
        min_df=2, # Note this
        max_df=0.8, # This, too
        binary=False,
        norm='l2',
        use_idf=True, # And this
        max_features=x)
    vect_n.append(vectorizer)
    matrix = vectorizer.fit_transform(training_books)
    X_train_Z = StandardScaler(with_mean=False).fit_transform(matrix)
    dict_key=str(x)
    matrix_n2[dict_key] = X_train_Z

In [62]:
scores2 = {} # Store cross-validation results in a dictionary
for matrix2 in matrix_n2: 
    scores2[matrix2] = cross_validate( # perform cross-validation
        MultinomialNB(), # classifier object
        matrix_n2[matrix2], # feature matrix
        y_train, # gold labels
        cv=10, #number of folds
        scoring=['accuracy','precision', 'recall', 'f1', 'f1_macro', 'f1_micro'] # scoring methods
    )

In [63]:
compare_scores(scores2)

#### Conclusion of testing MultiNomialNB() over different sized inputs closer to 17,500 features:
17,500 features still performed the best, tied with 19,000 features. **The best MultiNomialNB() classifier is with the default parameters and a matrix with 17,500 inputs at an F1 score of .9303.**

### Logistic Regression
Like MultinomialNB(), we also tested a variety of parameters, the relevant one being max_iter(), as seen below:

In [64]:
#fit log reg classifier
log_classifiers = {
    'log1':LogisticRegression(),
    'log2':LogisticRegression(max_iter = 1000),
    'log3':LogisticRegression(max_iter = 5000),
    'log4':LogisticRegression(max_iter = 10000),
    'log5':LogisticRegression(max_iter = 50000)
}

scores = {} # Store cross-validation results in a dictionary
for classifier in log_classifiers: 
    scores[classifier] = cross_validate( # perform cross-validation
        log_classifiers[classifier], # classifier object
        X_train_Z, # feature matrix
        y_train, # gold labels
        cv=10, #number of folds
        scoring=['accuracy','precision', 'recall', 'f1', 'f1_macro', 'f1_micro'] # scoring methods
    )
       
compare_scores(scores)

#### Conclusion of testing LogisticRegression() with different max_iter:
Like the naive bayes classifier, we found that toying with the parameters did nothing to change the F1 score or any of the scoring. We decided to stick with the default parameters in testing over different sized matrices.


In [65]:
scores = {} # Store cross-validation results in a dictionary
for matrix in matrix_n: 
    print(matrix_n[matrix].shape)
    scores[matrix] = cross_validate( # perform cross-validation
        LogisticRegression(), # classifier object
        matrix_n[matrix], # feature matrix
        y_train, # gold labels
        cv=10, #number of folds
        scoring=['accuracy','precision', 'recall', 'f1', 'f1_macro', 'f1_micro'] # scoring methods
    )

(80, 5000)
(80, 10000)
(80, 15000)
(80, 17500)
(80, 20000)
(80, 22500)
(80, 25000)
(80, 30000)
(80, 30376)


In [66]:
compare_scores(scores)

#### Conclusion of testing LogisticRegression() over different sized inputs: 
The default LogisticRegression() with a matrix of max_features=5000 performed the best across all scoring, with an F1 score at .9027. We decided to not test LogisticRegression() further because 5000 features seemed too little for our purpose, and refining this model would mean going below that number. We decided to go with the MultinomialNB() classifier instead, which perfomed about 3% better with an F1 score of about .9303.

## IV. Feature Importance

We wanted to see what specific features that were weighted most heavily in deciding if a model was detective or horror. Additionally, extracting these tokens is important to see if our model is using tokens that it shouldn't when doing classification, such as characters' names, articles, or punctuation. If the latter was the case, then we would go back and adjust our pre-processing for our vectorizer.

In [67]:
#source: https://stackoverflow.com/questions/50526898/how-to-get-feature-importance-in-naive-bayes
feat_impNB=MultinomialNB()
s=feat_impNB.fit(matrix_n2['17500'],y_train)

neg_class_prob_sorted = feat_impNB.feature_log_prob_[0, :].argsort()[::-1]
pos_class_prob_sorted = feat_impNB.feature_log_prob_[1, :].argsort()[::-1]


print("The most important words horror books:\n")
display(np.take(vectorizer.get_feature_names_out(), neg_class_prob_sorted[:50]))
print("\nThe most important words detective books:\n")
display(np.take(vectorizer.get_feature_names_out(), pos_class_prob_sorted[:50])) #??? detective=1

The most important words horror books:



array(['content', 'reality', 'closer', 'flesh', 'yield', 'choice',
       'attend', 'occur', 'struggled', 'whence', 'music', 'tender',
       'storm', 'solemn', 'mass', 'distress', 'dreaming', 'kissed',
       'flow', 'equal', 'trembled', 'gladly', 'wandering', 'travel',
       'reasonable', 'strongly', 'endless', 'glory', 'motion',
       'exhausted', 'provided', 'breathing', 'suffer', 'changes',
       'contrast', 'sleeping', 'composed', 'confusion', 'encourage',
       'considering', 'strain', 'wave', 'dreams', 'justified', 'animal',
       'tortured', 'branches', 'dignity', 'turns', 'treated'],
      dtype=object)


The most important words detective books:



array(['marry', 'morrow', 'suspicions', 'famous', 'aroused', 'suspicious',
       'warn', 'valuable', 'stolen', 'smoking', 'clue', 'profession',
       'conceal', 'arrested', 'hunting', 'interests', 'advised',
       'reasonable', 'impatiently', 'finds', 'objection', 'tragedy',
       'prefer', 'liberty', 'exclamation', 'picking', 'begged', 'mud',
       'visitors', 'tells', 'bending', 'absurd', 'gets', 'card',
       'considering', 'propose', 'disappearance', 'lunch', 'crushed',
       'consulted', 'investigation', 'gesture', 'employ', 'informed',
       'appreciate', 'cigar', 'detective', 'blame', 'replaced',
       'reputation'], dtype=object)

#### Observations of looking at the heaviest-weighted tokens for detective and horror
Overall, we can observe that the top tokens for each genre do not have overlap. Additionally, these tokens do not seem that strange for their genre, with detective using words like "suspicions" and "stolen" and horror using words like "distress" and "trembled."

## V. Testing/Predicting

Now that we fitted our best model, we can use the testing data of other novels and predict them as "horror" or "detective" using the classifier. It is important to note that we cannot check the accuracy of this fitting or score it because our testing_data does not consist of true labeled detective or horror.

In [68]:
vectorizer = TfidfVectorizer(
        encoding='utf-8',
        preprocessor=pre_proc,
        min_df=2, # Note this
        max_df=0.8, # This, too
        binary=False,
        norm='l2',
        use_idf=True, # And this
        max_features=17500)

In [69]:
X_test = vectorizer.fit_transform(testing_books)
y_test = MultinomialNB().fit(matrix_n['17500'], y_train).predict(X_test)

#### Detective=1, Horror=0

In [70]:
y_test

array([1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

### Testing books predicted Detective

Of the n=78 testing dataset, the following 68 novels were predicted as detective.

In [78]:
testing_data['y_pred']=y_test
print(len(testing_data[testing_data['y_pred']==1].title.values))
print("Book titles of other genres predicted detective:")
testing_data[testing_data['y_pred']==1].title.values

Number of books predicted detective:
68
Book titles of other genres predicted detective:


array(['A Round Dozen', 'A Sicillian Romance',
       'Adele Doring at Boarding-School', 'Agnes Grey',
       'An Old-Fashioned Girl', 'Anna Karenina', 'Don Quixote', 'Emma',
       'Flatland', 'Key Out of Time',
       "Little Men: Life at Plumfield With Jo's Boys", 'Little Women',
       'Mansfield Park', 'Mathilda', 'Micah Clarke', 'Middlemarch',
       'Mizora: A Prophecy', 'Moods', 'Mr. Standfast', 'Night and Day',
       "Nobody's Girl", 'Our Mutual Friend', 'Persuasion', 'Plague Ship',
       'Pride and Prejudice', 'Rainbow Valley', 'Sense and Sensibility',
       'Sense and Sensibility', 'Shirley', 'Silas Marner', 'Star Hunter',
       'Star of India', 'Storm over Warlock', 'Summer',
       'The Age of Innocence', 'The Beautiful and Damned', 'The Bell Jar',
       'The Best Made Plans', 'The Betrothed', 'The Colors of Space',
       'The Defiant Agents', 'The Disturbing Charm',
       'The Enchanted April', 'The Fall Of The Grand Sarrasin',
       'The Four Corners', 'The Garde

### Testing books predicted Horror

Of the n=78 testing dataset, only 10 were predicted horror

In [80]:
print("Number of books predicted horror:")
print(len(testing_data[testing_data['y_pred']==0].title.values))
print("Book titles of other genres predicted horror:")
testing_data[testing_data['y_pred']==0].title.values

Number of books predicted horror:
10
Book titles of other genres predicted horror:


array(['Anne of Green Gables', 'Black Amazon of Mars',
       'Chaplet Of Pearls', 'House of Mirth', 'The Deluge',
       'The Door Through Space',
       'The Importance of Being Earnest: A Trivial Comedy for Serious People by Oscar Wilde',
       'The Lances of Lynwood', 'The Luckiest Girl in the School',
       'The Mill On The Floss'], dtype=object)

### Comparison of ratio of detective and horror genres in the training vs. testing

63.75% of the novels in the training dataset were detective, and 87.18% of the novels in the testing dataset were predicted detective. The ratio of detective to horror is much larger in the testing compared to the training, which could mean that the majority of the testing novels are closer to detective than horror.

In [79]:
print("fraction of true detective novels in the training dataset:", len(training_data[training_data['detective']==True])/(len(training_books)))
print("fraction of novels of testing dataset predicted detective:", len(testing_data[testing_data['y_pred']==1].title.values)/len(testing_books))

fraction of true detective novels in the training dataset: 0.6375
fraction of novels of testing dataset predicted detective: 0.8717948717948718


## VI. Visualizations

This section is miscellaneous code used to prepare for visualizations and scatterplots that will be used in the Results and discussion section. This consists of using SVD on the entire corpus (using the vectorizer of max_features=17500) with 2 components to graph the novels in distance space in only 2 dimensions. The df books_order contains the metadata for all the books used in testing in training, adding the columns "y" to indicate if it was detective=1 or horror=0 (regardless if the label is true or predicted) and "shape" to indicate if the novel was part of the testing or training dataset.

In [74]:
all_books=training_books+testing_books
vis=vectorizer.fit_transform(all_books)
labels=np.hstack((y_train, y_test))

In [76]:
# source: mp2
coords_books = TruncatedSVD(n_components=2).fit_transform(vis)

In [75]:
training_data['shape']='train'
testing_data['shape']='test'
books_order=pd.concat([training_data, testing_data])
books_order['y']=labels
books_order['gender_author1']=books_order['gender_author1'].str.lower()
books_order

Unnamed: 0 check_1 check_2                                  title  year  \
0       tl566   hz542   ja532                                    813  1910   
1       gc386     NaN     NaN                A Strange Disappearance  1998   
2       nca28   tl566   stw43                     A Study in Scarlet  1887   
3      jc2739     NaN     NaN                            Agatha Webb  1899   
4       lcc82   yk499     NaN                               Carmilla  1872   
..        ...     ...     ...                                    ...   ...   
73     jc2739     NaN     NaN                  This Side of Paradise  1920   
74      vs339   thh55     NaN                  To Kill A Mockingbird  1960   
75      fhh26   gs542   tj256  Twenty Thousand Leagues Under the Sea  1870   
76      dgr73  jlp367   kg428                      Uncle Tom's Cabin  1852   
77     cl2264     NaN     NaN                          Voodoo Planet  1959   

   author1_surname author1_givenname author2_surname author2_givenname  \
0          Leblanc           Maurice             NaN               NaN   
1            Green    Anna Katharine             NaN               NaN   
2      Conan Doyle            Arthur             NaN               NaN   
3            Green    Anna Katharine             NaN               NaN   
4          Le_Fanu   Joseph Sheridan             NaN               NaN   
..             ...               ...             ...               ...   
73      Fitzgerald          F. Scott             NaN               NaN   
74             Lee            Harper             NaN               NaN   
75           Verne             Jules             NaN               NaN   
76           Stowe   Harriet Beecher             NaN               NaN   
77          Norton             Andre             NaN               NaN   

   gender_author1  ... tragedy children regency manners philosophical  \
0            male  ...   False    False   False   False         False   
1          female  ...   False    False   False   False         False   
2            male  ...   False    False   False   False         False   
3          female  ...   False    False   False   False         False   
4            male  ...   False    False   False   False         False   
..            ...  ...     ...      ...     ...     ...           ...   
73           male  ...   False    False   False   False         False   
74         female  ...   False    False   False   False         False   
75           male  ...   False    False   False   False         False   
76         female  ...   False    False   False   False         False   
77         female  ...   False    False   False   False         False   

    coming-of-age                                      filename  shape y_pred  \
0           False                               Leblanc_813.txt  train    NaN   
1           False  GreenAnnaKatharine_AStrangeDisappearance.txt  train    NaN   
2           False                ConanDoyle_AStudyInScarlet.txt  train    NaN   
3           False                          Green_AgathaWebb.txt  train    NaN   
4           False                                  Carmilla.txt  train    NaN   
..            ...                                           ...    ...    ...   
73           True             Fitzgerald_ThisSideOfParadise.txt   test    1.0   
74          False                    Lee_ToKillAMockingbird.txt   test    1.0   
75          False               Verne_TwentyThousandLeagues.txt   test    1.0   
76          False                     Stowe_UncleTom_sCabin.txt   test    1.0   
77          False                       Norton_VoodooPlanet.txt   test    1.0   

    y  
0   1  
1   1  
2   1  
3   1  
4   0  
.. ..  
73  1  
74  1  
75  1  
76  1  
77  1  

[158 rows x 37 columns]

# 4. Results and discussion (40 points)

In [77]:
# few paragraphs
# realistic - 3 results
# figure, table, accuracy table
# analyze each of the results, one paragraph at the end how they fit together
# group -- a bit more

# 5. Reflection (10 points)

Answering this research question was a lot harder than we expected. We thought choosing a classification problem would mean that the process would be really similar to MP 02, but found that that was not the case. When applying what we learned from lectures, such as standardization (which was not incorporated in MP 02), we found difficulty in knowing if we were better off doing something or not. Because we are classfying a group of novels that do not have the true labels of detective and horror like the training dataset, there is no way to know if our classifer was accurate on the testing dataset. For example, our z-scores did not have a mean-0 and we could not figure out how to get the standard deviation of a sparse matrix. Additionally, figuring out the feature importance was difficult because we hadn't learned how to calculate feature importance for text data or a naive bayes classifier, but dataframes with a small defined set of variables. Additionally, we did want to try the RandomForest classifier but we did not have the computational power to fit and try several random forest classifiers. Besides techinical and coding issues, we didn't achieve many interpretable results, perhaps because of the limitations of our corpus. When answering our research question, we used a general miscellaneous set of other genres, but in fact this testing dataset is not very diverse, and contains 73% female authors. Though we were able to define a set of recommendations at the end of our research, we knew that realistically many of the pairs of novels we found were not very similar to each other by reading the synopses of these novels. If we were to change our approach and access to a more diverse corpus, we would probably get a uniform sample of "other" novels, or make sure the other genres were represented uniformly (e.g. romance, science fiction). We would run the experiment over other combination pairs of genres besides detective and horror, like science fiction and historical, or romance and fantasy because our research question is not restricted to only the detective and horror genre. Alternatively, we could train a classifier that can predict more than a binary category, and that we could train a model that incorporates a training dataset and testing dataset with multiple genres each (and then we could have more faith in the classifier).

# 6. Resources consulted (0 points, but -5 if missing)

`pre_proc()`: MP 02, INFO 3350 <br>
`compare_scores()`: MP 02, INFO 3350<br>
SVD Scatterplot: MP 02, INFO 3350<br>
Lectures 12-15 INFO 3350 for reasoning in choosing a classifier and how to normalize and standardize <br>
Feature importance for sklearn Naive Bayes: https://stackoverflow.com/questions/50526898/how-to-get-feature-importance-in-naive-bayes<br>
MultinomialNB() sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.naive_bayes.MultinomialNB.html<br>
LogisticRegression() sklearn documentaton: https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html<br>
TfidfVectorizer() sklearn documentation: https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html<br>

# 7. Responsibility statement (0 points, but -5 if missing)
**See separate CMS assignment 'MP 03: Responsibility statement'.**